File: src/udemy_tutorials_py/udemy_tutorials_py/number_publisher.py
#!/usr/bin/env python3
"""Create a ROS2 node that publisheds a number to a topic.

    Publishers:
        number (std_msgs.msg.Int64): Publishes a number to the topic "number".
"""
import sys
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64


class NumberPublisher(Node):
    """ROS2 node that publishes a number to a topic."""
    def __init__(self):
        """Initialize the node and create a publisher."""
        super().__init__("number_publisher")
        self.__publisher = self.create_publisher(
            msg_type=Int64,
            topic="number",
            qos_profile=10,  # Queue size
        )
        # Publish a number to the topic every 1 second
        self.create_timer(
            timer_period_sec=1,
            callback=self.publish_number,
        )
        # Log a message to indicate that the node has been initialized
        self.logger.info("Number publisher node has been initialized.")

    @property
    def logger(self):
        """rclpy.logging.Logger: Node's logger."""
        return self.get_logger()

    def publish_number(self):
        """Publish a number to the topic."""
        number = 1
        # Publish the number to the topic. In this case, the number is 1, which
        # corresponds to the data field of the Int64 message.
        self.__publisher.publish(Int64(data=number))


def main():
    """Create and run the number publisher node."""
    rclpy.init(args=sys.argv)
    node = NumberPublisher()
    rclpy.spin(node)
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/udemy_tutorials_cpp/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"

/*
 * This example creates a publisher node that publishes a number to a topic.
 */


class NumberPublisher : public rclcpp::Node
{
    public:
        NumberPublisher() : Node("number_publisher")
        {
            // Creates a publisher that published a number to the topic "number"
            // The second argument is the size of the publisher queue
            publisher_ = this->create_publisher<std_msgs::msg::Int64>("number", 10);
            timer_ = this->create_wall_timer(
                std::chrono::seconds(1),
                // Callback
                std::bind(&NumberPublisher::publish_number, this)
            );
            RCLCPP_INFO(
                this->get_logger(),
                "NumberPublisher has been started"
            );
        } 

    private:
        void publish_number()
        {
            // Create a message to send
            auto msg = std_msgs::msg::Int64();
            msg.data = number_;
            publisher_->publish(msg);
        }
        int number_ = 1;
        rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char **argv)
{
    // Initialize ROS2
    rclcpp::init(argc, argv);
    // Create a NumberPublisher object
    auto node = std::make_shared<NumberPublisher>();
    // Spin the node
    rclcpp::spin(node);
    // Shut down ROS2 node
    rclcpp::shutdown();
    return 0;
}
